README.md
# Async Negotiate SSPI
**Credit brandond for the [requests_negotiate_sspi](https://github.com/brandond/requests-negotiate-sspi) package which this package is built off.**

## Background
This package takes the base code from requests_negotiate_sppi and adapts it to work within the context of httpx auth flows. It supports HTTP(S) and Websocket connections and plugs into both the [httpx](https://www.python-httpx.org/) and [websockets](https://websockets.readthedocs.io/en/stable/) API's. In order to plug into these API's, async_negotiate_sspi requires 2 dependencies...
* [httpx_extensions](https://github.com/newvicx/httpx_extensions): An async client built on top of httpx. Without it, HTTP(S) requests authenticating through Kerberos or NTLM using this package will fail. Its API is identical to httpx and has been fully tested
* [ws_auth](https://github.com/newvicx/httpx_extensions): An extension of the WebsocketClientProtocol class for handling httpx-like auth flows. No additional setup is needed with this and you can use the websockets API as you normally would

## Installation
You can install async_negotiate_sspi via pip

    pip install async-negotiate-sspi

## Usage

### HTTP(S)

    import asyncio
	from httpx_extensions import ExClient
	from async_negotiate_sspi import NegotiateAuth
	
	async def main():
		auth = NegotiateAuth()
		with ExClient(auth=auth) as client:
			response = await client.get("https://iis.contoso.com")
	
	asyncio.run(main())

### Websockets

    import websockets
	import ws_auth	# required to pass auth parameter
	from async_negotiate_sspi import NegotiateAuthWS
	
	async def main():
		auth = NegotiateAuthWS()
		with websockets.connect("wss://iis.contoso.com", auth=auth) as client:
			data = await client.recv()
	
	asyncio.run(main())

## Options

 - service (str): Kerberos Service type for remote Service Principal Name - Default: "HTTP"
 - username (str): Username - Default: None
 - domain (str): NT Domain name - Default: "."
 - host (str): Host name for Service Principal Name - Default: Extracted from request URI
 - delegate (bool): Indicates that the user's credentials are to be delegated to the server - Default: False

## Supports
* Python >=3.7

## Requires
* httpx_extensions>=0.1.1
* ws_auth>=0.1.0
* pywin32>=303
	



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## 0.1.3 (February 10, 2022)

### Added
* Websocket support through NegotiateAuthWS class
* Additional dependency: [ws_auth](https://github.com/newvicx/ws_auth) a websockets.WebsocketClientProtocol extension for handling httpx like auth flows

## 0.1.4 (March 3, 2022)
- Update requirements.txt
- Correct support to python >= 3.7 to align with Websockets

async_negotiate_sspi/__init__.py
from anyio.streams.tls import TLSAttribute, TLSStream
from httpcore.backends.asyncio import AsyncIOStream

from .negotiate import NegotiateAuth
from .negotiate_ws import NegotiateAuthWS

# Monkey patch AsnycIO and Trio Streams to expose peercert
orig_AsyncIOStream__init__ = AsyncIOStream.__init__

def new_AsyncIOStream__init__(self, *args, **kwargs):
    orig_AsyncIOStream__init__(self, *args, **kwargs)
    if isinstance(self._stream, TLSStream):
        self.peercert = self._stream.extra_attributes.get(TLSAttribute.peer_certificate_binary)()

AsyncIOStream.__init__ = new_AsyncIOStream__init__

try:
    from httpcore.backends.trio import TrioStream

    orig_TrioStream__init__ = TrioStream.__init__

    def new_TrioStream__init__(self, *args, **kwargs):
        orig_TrioStream__init__(self, *args, **kwargs)
        if isinstance(self._stream, TLSStream):
            self.peercert = self._stream.extra_attributes.get(TLSAttribute.peer_certificate_binary)()

    TrioStream.__init__ = new_TrioStream__init__
except (ImportError, ModuleNotFoundError):
    pass

async_negotiate_sspi/negotiate.py
import base64
import getpass
import hashlib
import logging
import socket
import struct
from typing import TypeVar, Union

import pywintypes
import sspi
import sspicon
import win32security

from httpcore.backends.asyncio import AsyncIOStream
try:
    from httpcore.backends.trio import TrioStream
except (ImportError, ModuleNotFoundError):
    from httpcore.backends.base import AsyncNetworkStream
    TrioStream = TypeVar("TrioStream", bound=AsyncNetworkStream)
from httpx import Auth, Request
from httpx._exceptions import HTTPError
from httpx_extensions import ResponseMixin


logger = logging.getLogger(__name__)


class NegotiateAuth(Auth):

    def __init__(
        self,
        service: str = "HTTP",
        username: str = None,
        domain: str = None,
        host: str = None,
        delegate: bool = False,
    ) -> None:

        """
        Create a new Negotiate auth handler

        Args:
            service: Kerberos Service type for remote Service Principal Name.
                Default: 'HTTP'
            username: If specified, will prompt for password
            domain: NT Domain name
                Default: '.' for local account.
            host: Host name for Service Principal Name.
                Default: Extracted from request URI
            delegate: Indicates that the user's credentials are to be delegated to the server.
                Default: False

        If username and password are not specified, the user's default credentials are used.
        This allows for single-sign-on to domain resources if the user is currently logged on
        with a domain account.
        """
        
        self._service = service
        self._auth_info = None
        if username is not None:
            password = getpass.getpass(f"Enter password for user account '{username}': ")
            domain = domain or '.'
            self._auth_info = (username, domain, password)
        self._delegate = delegate
        self._host = host

    def auth_flow(self, request: Request):

        """Perform Negotiate/NTLM authentication steps"""
        
        if request.headers.get("authorization") is not None:
            # If auth header provided do nothing
            yield request
            return
        
        response: ResponseMixin = yield request

        scheme = None
        allowed_schemes = ("Negotiate", "NTLM")
        for auth_header in response.headers.get_list("www-authenticate"):
            if auth_header.strip() in allowed_schemes:
                scheme = auth_header.strip()
                break
        if not scheme:
            # Server did not respond with Negotiate or NTLM
            return
        
        if self._host is None:
            self._host = request.url.host
            try:
                self._host = socket.getaddrinfo(self._host, None, 0, 0, 0, socket.AI_CANONNAME)[0][3]
                request.headers["host"] = self._host
            except socket.gaierror as err:
                logger.info('Skipping canonicalization of name %s due to error: %r', self._host, err)
        
        targetspn = '{}/{}'.format(self._service, self._host)

        # request mutual auth by default
        scflags = sspicon.ISC_REQ_MUTUAL_AUTH
        if self._delegate:
            scflags |= sspicon.ISC_REQ_DELEGATE

        pkg_info, clientauth, sec_buffer = self.init_sspi(scheme, targetspn, scflags)
        
        self.try_kerberos(request, response, scheme, pkg_info, clientauth, sec_buffer)

        # Get conn_id so next request can go out on same connection
        conn_id = response.extensions.get("conn_id")
        if not conn_id:
            logger.warning("No conn_id returned in response")
        else:
            logger.debug("Using conn_id %s", conn_id)
        request.extensions.update({"conn_id": conn_id})

        # send next request
        response: ResponseMixin = yield request
        
        if response.status_code != 401:
            # Kerberos may have succeeded; if so, finalize our auth context
            final = response.headers.get("www-authenticate")
            if final is not None:
                try:
                    # Sometimes Windows seems to forget to prepend 'Negotiate' to
                    # the success response, and we get just a bare chunk of base64
                    # token. Not sure why.
                    final = final.replace(scheme, '', 1).lstrip()
                    tokenbuf = win32security.PySecBufferType(pkg_info['MaxToken'], sspicon.SECBUFFER_TOKEN)
                    tokenbuf.Buffer = base64.b64decode(final.encode('ASCII'))
                    sec_buffer.append(tokenbuf)
                    error, auth = clientauth.authorize(sec_buffer)
                    logger.debug(
                        "Kerberos Authentication succeeded - error=%s authenticated=%s",
                        error, clientauth.authenticated
                    )
                except TypeError:
                    pass
            return
        
        # if kerberos failed, do NTLM
        self.try_ntlm(request, response, scheme, pkg_info, clientauth, sec_buffer)
        
        yield request

    def init_sspi(self, scheme: str, targetspn: str, scflags: int) -> Union[
        win32security.QuerySecurityPackageInfo,
        sspi.ClientAuth,
        win32security.PySecBufferDescType
    ]:
        """Set up SSPI connection structure"""

        pkg_info = win32security.QuerySecurityPackageInfo(scheme)
        clientauth = sspi.ClientAuth(
            scheme,
            targetspn=targetspn,
            auth_info=self._auth_info,
            scflags=scflags,
            datarep=sspicon.SECURITY_NETWORK_DREP
        )
        sec_buffer = win32security.PySecBufferDescType()
        return pkg_info, clientauth, sec_buffer

    def try_kerberos(
        self,
        request: Request,
        response: ResponseMixin,
        scheme: str,
        pkg_info: win32security.QuerySecurityPackageInfo,
        clientauth: sspi.ClientAuth,
        sec_buffer: win32security.PySecBufferDescType
    ) -> None:

        """Do channel binding hash step if SSL. If user has valid TGT, Kerberos should succeed on next request"""

        # Channel Binding Hash (aka Extended Protection for Authentication)
        # If this is a SSL connection, we need to hash the peer certificate,
        # prepend the RFC5929 channel binding type,
        # and stuff it into a SEC_CHANNEL_BINDINGS structure.
        # This should be sent along in the initial handshake or Kerberos auth will fail.
        stream: Union[AsyncIOStream, TrioStream] = response.extensions.get("network_stream")
        if hasattr(stream, "peercert") and stream.peercert is not None:
            peercert = stream.peercert
            md = hashlib.sha256()
            md.update(peercert)
            appdata = 'tls-server-end-point:'.encode('ASCII')+md.digest()
            logger.debug("Channel binding hash %s", appdata)
            cbtbuf = win32security.PySecBufferType(
                pkg_info['MaxToken'], sspicon.SECBUFFER_CHANNEL_BINDINGS
            )
            cbtbuf.Buffer = struct.pack('LLLLLLLL{}s'.format(len(appdata)), 0, 0, 0, 0, 0, 0, len(appdata), 32, appdata)
            sec_buffer.append(cbtbuf)
        
        # this is important for some web applications that store
        # authentication-related info in cookies
        if response.headers.get("set-cookie") is not None:
            request.headers["cookie"] = response.headers["set-cookie"]
        
        try:
            error, auth = clientauth.authorize(sec_buffer)
            auth_header = f"{scheme} {base64.b64encode(auth[0].Buffer).decode('ASCII')}"
            request.headers["authorization"] = auth_header
            logger.debug('Sending Initial Context Token - error=%s authenticated=%s', error, clientauth.authenticated)
        except pywintypes.error as err:
            logger.debug("Error in client auth: %r", repr(err))
            raise err
        return
    
    def try_ntlm(
        self,
        request: Request,
        response: ResponseMixin,
        scheme: str,
        pkg_info: win32security.QuerySecurityPackageInfo,
        clientauth: sspi.ClientAuth,
        sec_buffer: win32security.PySecBufferDescType
    ) -> None:

        """Get challenge from server and finalize auth context"""

        # Extract challenge message from server
        challenge_header = response.headers["www-authenticate"]
        challenge = [val[len(scheme)+1:] for val in challenge_header.split(', ') if scheme in val]
        if len(challenge) > 1:
            raise HTTPError(f"Did not get exactly one {scheme} challenge from server.")
        
        tokenbuf = win32security.PySecBufferType(pkg_info['MaxToken'], sspicon.SECBUFFER_TOKEN)
        tokenbuf.Buffer = base64.b64decode(challenge[0])
        sec_buffer.append(tokenbuf)
        logger.debug('Got Challenge Token (NTLM)')

        # Perform next authorization step
        try:
            error, auth = clientauth.authorize(sec_buffer)
            auth_header = f"{scheme} {base64.b64encode(auth[0].Buffer).decode('ASCII')}"
            request.headers["authorization"] = auth_header
            logger.debug('Sending response - error=%s authenticated=%s', error, clientauth.authenticated)
        except pywintypes.error as err:
            logger.debug("Error in client auth: %r", repr(err))
            raise err
        return

async_negotiate_sspi/negotiate_ws.py
import base64
import getpass
import hashlib
import logging
import socket
import struct
from typing import Union

import pywintypes
import sspi
import sspicon
import win32security
from httpx._exceptions import HTTPError
from ws_auth import Auth
from ws_auth.types import Request, Response


logger = logging.getLogger(__name__)


class NegotiateAuthWS(Auth):

    """Negotiate auth handler for WebsocketAuthProtocol"""

    def __init__(
        self,
        service: str = "HTTP",
        username: str = None,
        domain: str = None,
        host: str = None,
        delegate: bool = False,
    ) -> None:

        """
        Create a new Negotiate auth handler
        Args:
            service: Kerberos Service type for remote Service Principal Name.
                Default: 'HTTP'
            username: If specified, will prompt for password
            domain: NT Domain name
                Default: '.' for local account.
            host: Host name for Service Principal Name.
                Default: Extracted from request URI
            delegate: Indicates that the user's credentials are to be delegated to the server.
                Default: False
        If username and password are not specified, the user's default credentials are used.
        This allows for single-sign-on to domain resources if the user is currently logged on
        with a domain account.
        """
        
        self._service = service
        self._auth_info = None
        if username is not None:
            password = getpass.getpass(f"Enter password for user account '{username}': ")
            domain = domain or '.'
            self._auth_info = (username, domain, password)
        self._delegate = delegate
        self._host = host

    def auth_flow(self, request: Request):

        """Perform Negotiate/NTLM authentication steps"""
        
        request_uri = request[0]
        request_headers = request[1]
        if request_headers.get("Authorization") is not None:
            # If auth header provided do nothing
            yield request
            return
        
        response: Response = yield request
        response_headers = response[1]

        scheme = None
        allowed_schemes = ("Negotiate", "NTLM")
        auth_headers = response_headers.get("WWW-Authenticate")
        if not auth_headers:
            return
        for auth_header in auth_headers.split(','):
            if auth_header.strip() in allowed_schemes:
                scheme = auth_header.strip()
                break
        if not scheme:
            # Server did not respond with Negotiate or NTLM
            return
        
        if self._host is None:
            self._host = request_uri.host
            try:
                self._host = socket.getaddrinfo(self._host, None, 0, 0, 0, socket.AI_CANONNAME)[0][3]
                request_headers.pop("Host")
                request_headers.update({"Host": self._host})
            except socket.gaierror as err:
                logger.info('Skipping canonicalization of name %s due to error: %r', self._host, err)
        
        targetspn = '{}/{}'.format(self._service, self._host)

        # request mutual auth by default
        scflags = sspicon.ISC_REQ_MUTUAL_AUTH
        if self._delegate:
            scflags |= sspicon.ISC_REQ_DELEGATE

        pkg_info, clientauth, sec_buffer = self.init_sspi(scheme, targetspn, scflags)
        
        prepped_request = (request_uri, request_headers)
        next_request = self.try_kerberos(prepped_request, response, scheme, pkg_info, clientauth, sec_buffer)

        # send next request
        response: Response = yield next_request
        status_code = response[0]
        if status_code != 401:
            # Kerberos may have succeeded; if so, finalize our auth context
            response_headers = response[1]
            final = response_headers.get("WWW-Authenticate")
            if final is not None:
                try:
                    # Sometimes Windows seems to forget to prepend 'Negotiate' to
                    # the success response, and we get just a bare chunk of base64
                    # token. Not sure why.
                    final = final.replace(scheme, '', 1).lstrip()
                    tokenbuf = win32security.PySecBufferType(pkg_info['MaxToken'], sspicon.SECBUFFER_TOKEN)
                    tokenbuf.Buffer = base64.b64decode(final.encode('ASCII'))
                    sec_buffer.append(tokenbuf)
                    error, auth = clientauth.authorize(sec_buffer)
                    logger.debug(
                        "Kerberos Authentication succeeded - error=%s authenticated=%s",
                        error, clientauth.authenticated
                    )
                except TypeError:
                    pass
            return
        
        # if kerberos failed, do NTLM
        prepped_request = next_request
        next_request = self.try_ntlm(prepped_request, response, scheme, pkg_info, clientauth, sec_buffer)
        
        yield next_request

    def init_sspi(self, scheme: str, targetspn: str, scflags: int) -> Union[
        win32security.QuerySecurityPackageInfo,
        sspi.ClientAuth,
        win32security.PySecBufferDescType
    ]:
        """Set up SSPI connection structure"""

        pkg_info = win32security.QuerySecurityPackageInfo(scheme)
        clientauth = sspi.ClientAuth(
            scheme,
            targetspn=targetspn,
            auth_info=self._auth_info,
            scflags=scflags,
            datarep=sspicon.SECURITY_NETWORK_DREP
        )
        sec_buffer = win32security.PySecBufferDescType()
        return pkg_info, clientauth, sec_buffer

    def try_kerberos(
        self,
        request: Request,
        response: Response,
        scheme: str,
        pkg_info: win32security.QuerySecurityPackageInfo,
        clientauth: sspi.ClientAuth,
        sec_buffer: win32security.PySecBufferDescType
    ) -> None:

        """Do channel binding hash step if SSL. If user has valid TGT, Kerberos should succeed on next request"""

        request_uri = request[0]
        request_headers = request[1]
        response_headers = response[1]

        try:
            request_headers.pop("Authorization")
        except KeyError:
            pass
        # Channel Binding Hash (aka Extended Protection for Authentication)
        # If this is a SSL connection, we need to hash the peer certificate,
        # prepend the RFC5929 channel binding type,
        # and stuff it into a SEC_CHANNEL_BINDINGS structure.
        # This should be sent along in the initial handshake or Kerberos auth will fail.\
        socket = response[3]
        if socket is not None:
            peercert = socket.getpeercert(True)
        if peercert is not None:
            md = hashlib.sha256()
            md.update(peercert)
            appdata = 'tls-server-end-point:'.encode('ASCII')+md.digest()
            logger.debug("Channel binding hash %s", appdata)
            cbtbuf = win32security.PySecBufferType(
                pkg_info['MaxToken'], sspicon.SECBUFFER_CHANNEL_BINDINGS
            )
            cbtbuf.Buffer = struct.pack('LLLLLLLL{}s'.format(len(appdata)), 0, 0, 0, 0, 0, 0, len(appdata), 32, appdata)
            sec_buffer.append(cbtbuf)
        
        # this is important for some web applications that store
        # authentication-related info in cookies
        if response_headers.get("Set-Cookie") is not None:
            request_headers["Cookie"] = response_headers["Set-Cookie"]
        
        try:
            error, auth = clientauth.authorize(sec_buffer)
            auth_header = f"{scheme} {base64.b64encode(auth[0].Buffer).decode('ASCII')}"
            request_headers.update({"Authorization": auth_header})
            logger.debug('Sending Initial Context Token - error=%s authenticated=%s', error, clientauth.authenticated)
        except pywintypes.error as err:
            logger.debug("Error in client auth: %r", repr(err))
            raise err
        return request_uri, request_headers
    
    def try_ntlm(
        self,
        request: Request,
        response: Response,
        scheme: str,
        pkg_info: win32security.QuerySecurityPackageInfo,
        clientauth: sspi.ClientAuth,
        sec_buffer: win32security.PySecBufferDescType
    ) -> None:

        """Get challenge from server and finalize auth context"""

        request_uri = request[0]
        request_headers = request[1]
        response_headers = response[1]

        try:
            request_headers.pop("Authorization")
        except KeyError:
            pass
        # Extract challenge message from server
        challenge_header = response_headers["WWW-Authenticate"]
        challenge = [val[len(scheme)+1:] for val in challenge_header.split(', ') if scheme in val]
        if len(challenge) > 1:
            raise HTTPError(f"Did not get exactly one {scheme} challenge from server.")
        
        tokenbuf = win32security.PySecBufferType(pkg_info['MaxToken'], sspicon.SECBUFFER_TOKEN)
        tokenbuf.Buffer = base64.b64decode(challenge[0])
        sec_buffer.append(tokenbuf)
        logger.debug('Got Challenge Token (NTLM)')

        # Perform next authorization step
        try:
            error, auth = clientauth.authorize(sec_buffer)
            auth_header = f"{scheme} {base64.b64encode(auth[0].Buffer).decode('ASCII')}"
            request_headers.update({"Authorization": auth_header})
            logger.debug('Sending response - error=%s authenticated=%s', error, clientauth.authenticated)
        except pywintypes.error as err:
            logger.debug("Error in client auth: %r", repr(err))
            raise err
        return (request_uri, request_headers)